### PR TITLE
Extend pre-commit hooks to include Google Colab notebooks

### PR DESCRIPTION
**Pull request by ESKYoung**
# Summary

Google Colab notebooks have additional cell metadata that is **not** removed by the `pre-commit-jupyter` hook. Specifically, this cell metadata can include a user's full name, profile picture, and other personally-identifiable information.

To mitigate this, I have replaced `pre-commit-jupyter` with `nbstripout`, which has the added functionality of letting you specify what metadata to remove. I have tried to replicate the only extra change that `pre-commit-jupyter` makes (AFAIK), which is removing the `metadata.kernelspec` metadata; `nbstripout` is slightly different as it only deletes metadata, and does not replace entries like `pre-commit-jupyter`.

I also updated `detect-secrets` to the latest version, and updated `.secrets.baseline` and documentation accordingly. Due to contributors needing to have `detect-secrets` installed for scanning/auditing purposes, I've mandated the version in `requirements.txt`. This is not the case for `flake8`, so I have removed this from `requirements.txt`, since it is installed directly by `pre-commit`.

## Developer Testing

I've tested this using `pre-commit-jupyter`'s [test notebook](https://github.com/roy-ht/pre-commit-jupyter/blob/master/notebooks/test.ipynb) - note the way `nbstripout` keeps output cells is different.

I then used a pre-existing Google Colab notebook to check that the outputs were stripped, along with the kernel, and the user information. Note that keeping outputs is only possible via Jupyter, but Colab will preserve the tags/additional metadata (see updated documentation). After stripping outputs/metadata, I then loaded the notebook back in Google Colab to check it worked, and re-downloaded it again to ensure `keep_output` tags and metadata were retained.

`detect-secrets` was tested by deliberately trying to commit a secret-like variable in a tracked file on a throwaway branch (to trip the hook), scanning the baseline using the new command `detect-secrets scan --baseline .secrets.baseline`, then auditing the baseline. Once completed I was able to successfully commit a secret-like variable, as expected.

# Checklists

This pull/merge request meets the following requirements:

- [x] Code runs
- [x] Developments are **secure** and [**ethical**][data-ethics-framework]
- [x] You have made proportionate checks that the code works correctly
- [x] Test suite passes
- [x] [Minimum usable documentation][agilemodeling] written in the `docs` folder

Comments have been added below around the incomplete checks.

[agilemodeling]: http://agilemodeling.com/essays/documentLate.htm
[data-ethics-framework]: https://www.gov.uk/government/publications/data-ethics-framework

### CODE REVIEW

File: {{ cookiecutter.repo_name }}/docs/contributor_guide/pre_commit_hooks.md [hunk 1]
```diff
@@ -93,17 +94,34 @@ updating and auditing the `.secrets.baseline` file.
 
 ## Keeping specific Jupyter notebook outputs
 
+>  ℹ️ Currently (March 2020) there is no way to add tags and/or metadata to Google Colab notebooks. It's strongly
+> suggested that you download the Colab as a .ipynb file, and edit tags and/or metadata using Jupyter _before_
+> committing the code if you want to keep some outputs.
+
```
Comment (exfalsoquodlibet) [inline {{ cookiecutter.repo_name }}/docs/contributor_guide/pre_commit_hooks.md:100]: Just a thought but it is a bit weird to find information about Google Colab as first thing first in a section about Jupyter notebook outputs.

I'd move it at the end of this section.

File: {{ cookiecutter.repo_name }}/docs/contributor_guide/pre_commit_hooks.md [hunk 2]
```diff
@@ -93,17 +94,34 @@ updating and auditing the `.secrets.baseline` file.
...
+(View > Cell Toolbar > Tags; View > Cell Toolbar > Edit Metadata). For the tags approach, enter `keep_output` in the
+text field for each desired cell, and press the "Add tag" button. For the metadata approach, press the "Edit Metadata"
+button on each desired cell, and edit the metadata to look like this:
 
-```julia
-# [keep_output]
+```json
+{
+  "keep_output": true
+}
```
Comment (exfalsoquodlibet) [inline {{ cookiecutter.repo_name }}/docs/contributor_guide/pre_commit_hooks.md:116]: Very helpful!

**Review by exfalsoquodlibet — COMMENTED** [review]
Thanks @ESKYoung.

I have verified that the required hooks are correctly triggered when committing Jupyter notebooks and Google colab notebook with outputs. I also verified that no hook is trigged when exceptions are explicitly encoded in the notebook’s tags or metadata cells.

As for update to detect-secrete.  

The [test notebook](https://github.com/roy-ht/pre-commit-jupyter/blob/master/notebooks/test.ipynb) triggered an `ERROR: Potential secrets about to be committed to git repo!`  It is picking up on the image’s “code” :D

> Secret Type: Base64 High Entropy String
> Location:    notebooks/test.ipynb:51
> Possible mitigations:
> - For information about putting your secrets in a safer place, please ask in #security
>  - Mark false positives with an inline `pragma: allowlist secret` comment

I think this is automatically generated by the Yelp module, While we happen to have a `#security` slack channel at GDS, such slack channel does not exist in the x-gov Slack channel and we cannot assume users will have access to a `#security` channel. 

Also, the second instruction on how to deal with a false positive is different from the one we have included in the `{{ cookiecutter.repo_name }}/docs/contributor_guide/pre_commit_hooks.md`.

I was trying to find a way to delete and change that message but couldn't find it.

The only other comment I have is that, in Google Colab, a contributor could potentially save the notebook directly from Colab too Github (if allowed by their organisation). If this is the case, the notebook is outside the scope of the pre-commit hooks. Shall we maybe add a simple warning?  Like: 
> WARNING: no pre-commit hooks check will be run on Google Colab saved directly to GitHub. 
It is obvious but...

File: {{ cookiecutter.repo_name }}/docs/contributor_guide/pre_commit_hooks.md (hunk 1 above)
Comment (ESKYoung) [reply {{ cookiecutter.repo_name }}/docs/contributor_guide/pre_commit_hooks.md]: Resolved in commit [`6164ea5`](https://github.com/ukgovdatascience/govcookiecutter/pull/24/commits/6164ea5f1b6a3d1f0887ae7e20a8aff08b2b9a7e).

**Comment by ESKYoung** [comment]
@exfalsoquodlibet thanks for the review! I've made the changes requested - added more clarification on the false positive instructions (commit [`6164ea5`](https://github.com/ukgovdatascience/govcookiecutter/pull/24/commits/6164ea5f1b6a3d1f0887ae7e20a8aff08b2b9a7e)) and a warning about pushing Colab notebooks directly to GitHub (commit [`fd043f6`](https://github.com/ukgovdatascience/govcookiecutter/pull/24/commits/fd043f6cf844c2a4edc8432241760434aa305704)).

> I think this is automatically generated by the Yelp module, While we happen to have a #security slack channel at GDS, such slack channel does not exist in the x-gov Slack channel and we cannot assume users will have access to a #security channel.

I think there's an ongoing issue with this (see [here](https://github.com/Yelp/detect-secrets/issues/436)). I suggest we leave it for now, and keep an eye on that issue for changes!

**Review by exfalsoquodlibet — APPROVED** [review]
Thanks for the changes @ESKYoung. Happy for this PR to merged to `main.

**Pull request merged**